compiler/noirc_evaluator/src/ssa/parser.rs

```diff
@@ -0,0 +1,830 @@
+        } else if self.eat_keyword(Keyword::Fold)? {
+            Ok(InlineType::Fold)
+        } else if self.eat_keyword(Keyword::NoPredicates)? {
+            Ok(InlineType::NoPredicates)
+        } else {
+            self.expected_one_of_tokens(&[
+                Token::Keyword(Keyword::Inline),
+                Token::Keyword(Keyword::InlineAlways),
+                Token::Keyword(Keyword::Fold),
+                Token::Keyword(Keyword::NoPredicates),
+            ])
+        }
+    }
+
+    fn parse_blocks(&mut self) -> ParseResult<Vec<ParsedBlock>> {
+        let mut blocks = Vec::new();
+        while !self.at(Token::RightBrace) {
+            let block = self.parse_block()?;
+            blocks.push(block);
+        }
+        Ok(blocks)
+    }
+
+    fn parse_block(&mut self) -> ParseResult<ParsedBlock> {
+        let name = self.eat_ident_or_error()?;
+        self.eat_or_error(Token::LeftParen)?;
+
+        let mut parameters = Vec::new();
+        while !self.at(Token::RightParen) {
+            parameters.push(self.parse_parameter()?);
+            if !self.eat(Token::Comma)? {
+                break;
+            }
+        }
+
+        self.eat_or_error(Token::RightParen)?;
+        self.eat_or_error(Token::Colon)?;
+
+        let instructions = self.parse_instructions()?;
+        let terminator = self.parse_terminator()?;
+        Ok(ParsedBlock { name, parameters, instructions, terminator })
+    }
+
+    fn parse_parameter(&mut self) -> ParseResult<ParsedParameter> {
+        let identifier = self.eat_identifier_or_error()?;
+        self.eat_or_error(Token::Colon)?;
+        let typ = self.parse_type()?;
+        Ok(ParsedParameter { identifier, typ })
+    }
+
+    fn parse_instructions(&mut self) -> ParseResult<Vec<ParsedInstruction>> {
+        let mut instructions = Vec::new();
+        while let Some(instruction) = self.parse_instruction()? {
+            instructions.push(instruction);
+        }
+        Ok(instructions)
+    }
+
+    fn parse_instruction(&mut self) -> ParseResult<Option<ParsedInstruction>> {
+        if let Some(instruction) = self.parse_call()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_constrain()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_decrement_rc()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_enable_side_effects()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_increment_rc()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_range_check()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(instruction) = self.parse_store()? {
+            return Ok(Some(instruction));
+        }
+
+        if let Some(target) = self.eat_identifier()? {
```

Comment:
Would perhaps `parse_assignment` be an adequate name for this `if` and its body? It's quite long, all the others are one liners. I was just looking for when `Ok(None)` is returned and it would make it easier to see that it's not from within anywhere in this.

Reply:
Good idea! I didn't do it because to extract each instruction into a method it couldn't work because `target` would have moved, but at least extracting `parse_assignment` for that entire section makes things a bit simpler.

I'm thinking the assignment instruction could be split into `target` and `rest` so we can parse rest without having to move `target`, but I don't know if it's worth it. What do you think?

Reply:
I don't fully understand what "moving the target" means, but I think the following would be the nicest pattern in the context of what is above it:

```rust
        if let Some(instruction) = self.parse_store()? {
            return Ok(Some(instruction));
        }

        if let Some(instruction) = self.parse_assignment()? {
            return Ok(Some(instruction));
        }

        Ok(None)
```

If there were multiple options to parse `target` then it would be different, but it looks like `parse_assignment` is the only one. If it was like this:
```
if let Some(target) = self.eat_identifier()? {
   if let Some(ins) = self.parse_foo(target)? { ... }
   if let Some(ins) = self.parse_bar(target)? { ... }
   return Err(self.expected_foo_or_bar());
}
```
then I think it think the placement of `return self.expected_instruction_or_terminator();` would be more justified, but the way it is it might as well be in `parse_assignment`.

Reply:
Done!

With "moving the target" I meant that right now `parse_assignment` is still a bit long. It could be split into helper functions but `self.parse_foo(target)?` won't work because `target` moves to `parse_foo` and then we can't call `parse_bar`.